Feature/support channel selection certification flow capi commands

morantr opened the pull request:
Add certification flag to the controller configuration file with bml setters and getters.
When this flag is set, the channel selection flow is blocked, until the controller will receive the relevant WFA-CA command to continue the flow.
#121

morantr committed:
common: btlvf: Add certification mode request/response messages

Add certification mode setter and getter messages in the BML messages.
Add the autogenerated code of these new messages.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Add certification mode to controller

Add certification mode flag to controllers configuration file.
Add to controller DB api to change this flag.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Change certification mode from bml

Add flow which allows getting/set controllers certification mode from
BML.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Hold channel selection flow on certification mode

1. Allocate certification buffer on son_master init function if the
certification flag is set.

2. When certification mode is enabled, disable automatic sending of
channel preference query upon receiving M1.

3. Disable automatic response of channel selection request after
receiving channel preference report, and saving the CMDU data on
the certification buffer instead.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: cli: fix wrong parsing of quotations marked commands

CLI parses command into tokens on space delimiter.
If the CLI get a command that its argument is in quotations marks,
then it should read it as one token and not try to split it to tokens.
For example: bml_some_command "arg1 arg2 arg3"
because of the quation marks, there will be only one token:
token1='arg1 arg2 arg3'
instead of token1='arg1', token2='arg2', token3='arg3'

When BML receives quotations marked command, it dropped it if it
had spaces on it, because it found out that there are more arguments
that there should be.
To solve this, moved the check for quotations marked command to
before the number of argument qualifier, and override the number
of arguments to 1.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
framework: Add load function to CmduTx

Add load function to CmduMessageTx that allows building CmduTx
from an existing buffer.
Removed as well m_parse checker on CmduMessageTx finalize function.
It was there because of historical reasons when cmdu_tx and cmdu_rx
where the same object.
This check does not really give something and fails finalization of
cmdu_tx created from an existing buffer.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Add infrastructure to WFA-CA module

1. Add forbidden characters checker.
2. Add command parser.
3. Remove reply_invalid_error() function and make the function
reply() to support all of WFA-CA statuses.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Add support to WFA-CA 'DEV_RESET_DEFAULT' command

Add parsing and checking of command arguments, and send a reply.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Add support to WFA-CA 'DEV_SET_CONFIG' command

Add parsing and checking of command arguments, and send a dummy reply.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Add support to WFA-CA 'DEV_SEND_1905' command

Add parsing and checking of command arguments, and send a reply.
Add function to WFA-CA module that creates CMDU, and may create it
from a certification buffer that contained a preprepared CMDU buffer.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

arnout commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -3044,6 +3044,25 @@ void db::lock() { db_mutex.lock(); }
+void db::setting_certification_mode(bool en)
+{
+    if (!en) {
+        remove_certification_tx_buffer();
```
Why is all this complicated certification_tx_buffer handling needed? Why don't we just follow the same pattern as e.g. dfs_reentry?

As far as I can see, it is not really used anywhere. It is filled from the CMDU_TX but otherwise ignored, it seems.

Whenever something is not immediately obvious, it should be explained either in a comment or (usualy) in the commit message. The commit message that adds the certification_tx_buffer doesn't even mention this, it just talks about the flag and the DB API change.

morantr replied:
We could choose to declare a static buffer or allocate and deallocate it according to the certification flag. I asked Tomer and he preferred to allocate and deallocate it.
I don't understand why it is not obvious to you.

tomereli replied:
It is not obvious since the obvious way would have been to add what we need to the DB when we get the request and not save the RX CMDU at all. You said it will make things complicated, it will help if you can explain why the DB approach cannot be used

arnout commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -463,6 +464,24 @@ class db {
+    std::shared_ptr<uint8_t> &allocate_certification_tx_buffer()
+    {
+        return (certification_tx_buffer =
+                    std::shared_ptr<uint8_t>(new uint8_t[beerocks::message::MESSAGE_BUFFER_LENGTH],
```
I don't understand why this buffer is needed, and I *defnitely* don't understand why it needs to be MESSAGE_BUFFER_LENGTH instead of 1 byte.

morantr replied:
We use the same buffer length of the controller tx_buffer, which is actually declared in beerocks_socket_thread. The reason for the size is because we copy the cmdu_tx buffer (tx_buffer) to the certification buffer when the certification flag is set.

arnout commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -59,6 +59,7 @@ class db {
         bool load_front_measurements;
         bool load_health_check;
         bool load_monitor_on_vaps;
+        bool certification_mode;
```
I'm not sure what the difference is between sDbMasterConfig and sDBMasterSettings, but I feel it is more appropriate to put this in Settings, together with things like health_check and rdkb_extensions.

morantr replied:
sDBMasterSettings if for dynamic configuration which we do logic 'AND' with the configuration in sDbMasterConfig.
Since we don't do it and override the configuration in the config file, it is defined in sDBMasterSettings.

arnout commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -695,6 +695,24 @@ int bml_internal::process_cmdu_header(cmdu_vs_action_header_t beerocks_header,
+        } break;
+        case beerocks_message::ACTION_BML_GET_CERTIFICATION_MODE_RESPONSE: {
+            auto response =
+                cmdu_rx.addClass<beerocks_message::cACTION_BML_GET_DFS_REENTRY_RESPONSE>();
```
This doesn't seem right... Copy & paste error?

morantr replied:
yes. fixed.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -506,6 +512,10 @@ bool master_thread::handle_cmdu_1905_autoconfiguration_WSC(Socket *sd,
     }
 
+    if (database.setting_certification_mode()) {
+        return true;
```
I think it's better to enclose the trigger itself in the condition rather than returning here. We can return because the channel selection trigger *happens* to be last thing, but if we ever add something after it, it's not so obvious that it's gated by the certification mode.

Also, I don't understand why the chennel preference query should not be sent in certification mode. The "why" should be part of the commit message. I *guess* it's because the test specification doesn't talk about the channel preference query, and conversely it does have an explicit CAPI command to trigger channel preference query. However, there is nothing that says the DUT should not send any messages that are not specified in the test plan. So I don't think this is needed at all.

morantr replied:
Enclosed it.
Regarding the rest of your comment - I was explained in the meeting.

arnout commented on controller/src/beerocks/cli/beerocks_cli.cpp:
```diff
@@ -91,9 +91,18 @@ bool cli::parsedCommandArgumentsLinux(std::string line)
     const commandInfo &info = commandIt->second;
     commandNumOfArgs        = commandTokens.size() - 1;
 
+    int i                = 0;
```
This declaration shouldn't have moved up with the rest, the variable is only used at the end.

morantr replied:
fixed

arnout commented on controller/src/beerocks/cli/beerocks_cli.cpp:
```diff
@@ -91,9 +91,18 @@ bool cli::parsedCommandArgumentsLinux(std::string line)
+    bool late_processing = false;
+    auto first_arg       = std::next(commandTokens.begin());
+    if (first_arg != commandTokens.end() && (*first_arg)[0] == '"') {
+        late_processing  = true;
```
This is a pretty stupid way of handling quotes - but OK, it was already like that, and this simple change indeed fixes it. Simple is better!

morantr replied:
Why do you think it is stupid? how would you do it instead?

arnout replied:
Sorry, stupid was the wrong word, I meant "naive", because it only supports putting all arguments together between quotes, i.e.
```
command "foo bar baz"
```
and not something like
```
command "foo bar" baz
comman2 foo "bar baz"
```

morantr replied:
Yes, there is no other command the has spaces in a single argument. The WFA-CA command is the only command that has it and also, all the text it the command should be read as a single argument because it is not processed in the cli.
So yeah it was the simple naive solution because we didn't want anything more back then.

arnout reviewed (changes requested):
I only reviewed up to "controller: cli: fix wrong parsing of quotations marked commands" - I'll handle the rest later.

arnout commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -11,8 +11,32 @@
 
 #include <algorithm>
 
+#define FORBBIDEN_CHARS "$#&*()[]{};:,/\\<>?|`~=+"
```
FORBBIDEN -> FORBIDDEN

I personally prefer `static const std::string g_forbidden_chars("...");` (nitpick)

morantr replied:
fix the typo.
Regarding the macro - It is only used in one place... it doesn't really matter.

tomereli replied:
I also prefer static const.

arnout commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -11,8 +11,32 @@
+
+static std::string check_forbidden_chars(const std::string &str)
+{
+    std::string forbbiden_chars_found;
```
Same here: forbbiden -> forbidden

morantr replied:
fixed

arnout commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -53,7 +77,7 @@ wfa_ca::eWfaCaCommand wfa_ca::wfa_ca_command_from_string(std::string command)
                                  eWfaCaStatus status, const std::string &description)
 {
-    if (status != eWfaCaStatus::INVALID || status != eWfaCaStatus::ERROR) {
+    if (status != eWfaCaStatus::INVALID && status != eWfaCaStatus::ERROR) {
```
Please, do fixes like this in a separate commit, with an explanation of what they fix.

morantr replied:
I didn't think it was a significant change for a separate commit.
And since this module is currently in development, it felt natural to me to push it as part of this commit.

tomereli replied:
still, take this as an area for improvement for next time - splitting to commits especially if they are not related to the actual change helps for a better code review process. You should keep in mind that code review is essential and benefitial especially to the one who is doing the change, so thriving for a clear and well split commit is a requirement, not a recommendation.

arnout commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -11,8 +11,32 @@
+    return g_token_idx == g_cmd_tokens_vec.size();
+    LOG(DEBUG) << "Tokens_num=" << int(g_cmd_tokens_vec.size())
+               << ", g_token_idx=" << int(g_token_idx);
+}
```
Why are all of the above globals instead of members of wfa_ca? I'm not against globals per se, but there should be a reason for it. Especially because we have a multithreaded application, it's something to be careful with.

Reviewing how this is used, I believe instead they should be passed as arguments to parse_param, as I'll explain below.

morantr replied:
It is global in a CPP file that runs on a single thread.
I think it is cleaner how I did it.

arnout replied:
I strongly disagree that it is cleaner. There is simply no reason to use globals in this case, except to avoid having to pass an argument to a function.

morantr replied:
fine I changed it. But I really believe that what I did looks better :\

tomereli replied:
I strongly agree with Arnout on this one. Globals are evil and should be avoided if not absolutely necessary 😉

arnout commented on controller/src/beerocks/master/wfa_ca.h:
```diff
@@ -40,8 +42,15 @@ class wfa_ca {
     static void reply_invalid_error(Socket *sd, ieee1905_1::CmduMessageTx &cmdu_tx,
                                     eWfaCaStatus status, const std::string &description);
 
+    static bool reply_running(Socket *sd, ieee1905_1::CmduMessageTx &cmdu_tx);
```
Let me tell you that I really like how you introduced the basic wfa_ca class in the previous PR, and refactor new functionality into it every time you can use it. IMO that is much more useful than trying to think of all the needed functionality up front.

morantr replied:
thanks :)

arnout commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -64,6 +88,22 @@ void wfa_ca::reply_invalid_error(Socket *sd, ieee1905_1::CmduMessageTx &cmdu_tx,
+
+void wfa_ca::reply_complete(Socket *sd, ieee1905_1::CmduMessageTx &cmdu_tx,
+                            const std::string &description)
+{
```
Since the wfa_ca replies are always of the form `status,STATUS[,reply]`, I think it would be more useful to add the body of this function to reply() itself, and give reply() the signature

```
bool reply(Socket *sd, ieee1905_1::CmduMessageTx &cmdu_tx, eWfaCaStatus status, const std::string &description = "");
```

With that, the reply_invalid() function is probably no longer useful, you can just call reply() directly. Maybe even reply_running and reply_complete are not so useful (the just save setting the eWfaCaStatus).

(nitpick)

morantr replied:
changed

arnout commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -95,20 +175,32 @@ void wfa_ca::handle_wfa_ca_message(
+    }
+
+    g_cmd_tokens_vec = string_utils::str_split(command_str, ',');
+    g_token_idx      = 0;
```
I think this is better as:

```
auto cmd_tokens_vec = string_utils::str_split(command_str, ',');

if (cmd_tokens_vec.empty()) {... error ...}

std::string command_type_str = cmd_tokens_vec.begin();
cmd_tokens_vec.erase(cmd_tokens_vec.begin());
...
parse_params(cmd_tokens_vec);
```

arnout replied:
I really believe this is a lot cleaner and easier to understand than the explicit indexing in a global variable, even if it requires calling std::next inside the loop inside the parse_params function. The current code basically re-implements a standard iterator.

morantr replied:
fine I changed it. But I really believe that what I did looks better :\

arnout commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -83,6 +123,46 @@ bool wfa_ca::reply(Socket *sd, ieee1905_1::CmduMessageTx &cmdu_tx, const std::st
     return message_com::send_cmdu(sd, cmdu_tx);
 }
 
+bool wfa_ca::parse_params(std::unordered_map<std::string, std::string> &params,
```
It is very weird to mix the introduction of this function in a commit that also does 4 other things, but does not actually call this function.

Please please please do 1 thing per commit. I spend too much time reviewing because of this. When you have just implemented it, you understand what the code is doing, so it is easy to see which parts belong together and put them together in one commit. When reviewing, you have no idea what is going on, so trying to see if what has changed is indeed the right thing to do takes a lot more mental energy and thus time.

Remember, the point of reviewing is to check if the change what was done is *the right thing* and *for the right reasons*. The first part is easier to validate when the commit is properly split up. The second part is easier to validate when there is a commit message that explains the reasons. If the reviewer has to make a guess at all of this, it will take a lot more time.

morantr replied:
Yeah, I try to make infrastructure changes that will be used in the follow-up commit -preparation commit.
I guess I understand why it is more default to review change that it is not used.
In CONTRIBUTING.md it is said:

> "Since commits should be split up, there will be many commits that are not useful on their own, but just prepare for another commit. In that case, the preparatory commits should refer to what they prepare for."

I see now that I forgot to mention in the commit message that it is a preparatory commit.
Next time :)

arnout commented on controller/src/beerocks/master/wfa_ca.h:
```diff
@@ -40,8 +42,15 @@ class wfa_ca {
     static bool reply(Socket *sd, ieee1905_1::CmduMessageTx &cmdu_tx,
                       const std::string &reply_string);
+
+    static bool parse_params(std::unordered_map<std::string, std::string> &params,
```
I haven't mentioned this in earlier reviews, but we agreed that all new functions and classes should have a doxygen comment.

arnout commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -83,6 +123,46 @@ bool wfa_ca::reply(Socket *sd, ieee1905_1::CmduMessageTx &cmdu_tx, const std::st
+    }
+
+    size_t mandatory_params_cnt = 0;
+    while (!finished_tokens()) {
```
This could just be
```
for (auto param_name : cmd_tokens_vec)
```
(assuming that that is passed as an argument to the function, obviously).

This removes the need for get_next_token() and finished_tokens() and all global variables.

morantr replied:
No, it doesn't fit here, because not all the elements in the vector are "param_name".
After "param_name" comes the param_value, and I need both of them in the same iteration.
I guess I could do it with for loop with an iterator and use std::next() to update the iterator, but I think it is more clear/clean the way I did it.

arnout commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -83,6 +123,46 @@ bool wfa_ca::reply(Socket *sd, ieee1905_1::CmduMessageTx &cmdu_tx, const std::st
+    while (!finished_tokens()) {
+        std::string param_name = get_next_token();
+        std::transform(param_name.begin(), param_name.end(), param_name.begin(), ::tolower);
+        if (params.find(param_name) != params.end()) {
```
I guess here you meant mandatory_params instead of params.

I don't think that will work well. I think it's better to pass mandatory_params explicitly as an input argument. Then you can make a copy of it (implicit because you don't pass it by reference) and just delete the ones that you find. That also solves the issue that counting is incorrect if the same parameter occurs twice.

morantr replied:
It works well, I tested it.
Your way will work as well. The only difference is how many arguments you have to the function. Mine has one less.
In my implementation, I can also delete mandatory params from the list and at the end of the function, check if the list is empty, but the counter is good too.
I don't see a reason to change it.

arnout replied:
True, you could delete it from the list, I didn't think of that.

Anyway, I should have put (nitpick) here.

arnout commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -83,6 +123,46 @@ bool wfa_ca::reply(Socket *sd, ieee1905_1::CmduMessageTx &cmdu_tx, const std::st
+            return false;
+        }
+
+        params[param_name] = get_next_token();
```
So here you can use `param_name.next()`.

arnout commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -83,6 +123,46 @@ bool wfa_ca::reply(Socket *sd, ieee1905_1::CmduMessageTx &cmdu_tx, const std::st
+            err_string += "'" + param_name + "', ";
+        }
+    }
+
```
Since the caller always has the `LOG(ERROR)` and `reply_invalid` calls, it would make sense to those here instead.

morantr replied:
I'm sorry, can you please rephrase that - "it would make sense to those here instead"

arnout replied:
It would make sense to add those here. I.e., instead of sending the reply and doing the log in the caller, do it directly here.

However, I realise now that sending a reply from this function wouldn't be a good idea, because the other replies are sent from the caller. It is indeed not a good idea to sometimes send the reply from one function and sometimes from another. So it's fine as it is.

arnout reviewed (changes requested):
Still not done... Now up to "controller: Add infrastructure to WFA-CA module"

This is a really big PR, it would make sense to split it up into two PRs.

morantr commented:
@arnout, After I squash all the fixup I'll fix the commit messages.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -513,14 +513,12 @@ bool master_thread::handle_cmdu_1905_autoconfiguration_WSC(Socket *sd,
+        }
     }
 
     son_actions::send_cmdu_to_agent(sd, cmdu_tx);
```
I think this has to go inside the condition as well, no?

morantr replied:
YesI missed it somehow. Thanks.

arnout commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -17,36 +17,38 @@
+/**
+ * @brief: Checks for forbidden characters inside WFA-CA command
+ * 
+ * @param[in str: WFA-CA command .
```
Missing closing brace: `@param[in]`. Also, I don't think there should be a colon after it. Doxygen will take the first word that comes after the param command as the parameter name.

morantr replied:
fixed

arnout commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -17,36 +17,38 @@
+ * @brief: Checks for forbidden characters inside WFA-CA command
+ * 
+ * @param[in str: WFA-CA command .
+ * @return std::string: String containing all the forbidden characters the have been found.
```
I don't think we should mention the type of the return value.

(nitpick)

arnout commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -65,6 +67,12 @@ const std::string wfa_ca::wfa_ca_status_to_string(eWfaCaStatus status)
+ * @brief: Convert string of WFA-CA command into enum
+ * 
+ * @param[in] command: String of WFA-CA command.
+ * @return wfa_ca::eWfaCaCommand: Command enum.
```
Here it is useful to mention that WFA_CA_COMMAND_MAX will be returned if the command is not recognise. The "exceptions" are often the main use of the doxygen comment. The other very useful thing is what constraints exist on the input arguments (in this case nothing special so nothing needs to be mentioned).

arnout commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -133,36 +131,50 @@ bool wfa_ca::reply(Socket *sd, ieee1905_1::CmduMessageTx &cmdu_tx, const std::st
+ * @brief: Parse the strings vector which containing parameter names, and values into unordered map
+ * which will contain the parameter name as a map key, and paramaeter value as map value.
+ * The function receives as an input unordered map of parameters that are mandatory, which will
+ * will include non-mandatory parameters as output.
```
Double "will".

BTW I'm really happy with this documentation, I wish the rest of the code also had it :-)

arnout commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -133,36 +131,50 @@ bool wfa_ca::reply(Socket *sd, ieee1905_1::CmduMessageTx &cmdu_tx, const std::st
+ * will include non-mandatory parameters as output.
+ * 
+ * @param[in] command_tokens: A vector containing commands parameters name and values.
+ * @param[in/out] params: An unordered map containing mandatory parameters on input, and parsed 
```
Input-output parameters are specified as `[in,out]` (see http://doxygen.nl/manual/commands.html#cmdparam). (nitpick, because I think doxygen will anyway just print it as is).

morantr replied:
fixed

arnout commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -133,36 +131,50 @@ bool wfa_ca::reply(Socket *sd, ieee1905_1::CmduMessageTx &cmdu_tx, const std::st
-    while (!finished_tokens()) {
-        std::string param_name = get_next_token();
+    for (auto it = std::next(command_tokens.begin()); it != command_tokens.end(); it++) {
+        std::string param_name = *it;
```
For my understanding (I don't have that much experience with C++...): could we use `auto` here as well? Or would that have the undesired effect of creating a reference instead of a copy?

In some other review I made the remark "we generally use auto", but now I'm not so sure because it really does help me to see that it is a string explicitly, without having to go through the iterator type and everything to find out.

morantr replied:
if I want it to be a reference I would do:
`auto& param name = *it;`

But yes, we generally want to use auto, but in some cases, you would prefer to explicitly write the type because it helps you to understand the code better.

arnout commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -203,28 +203,123 @@ void wfa_ca::handle_wfa_ca_message(
+        * "devrole"     | string ("controller"/"agent") | Device role    
+        * "name"        | string                        | Device name    
+        * "program"     | string ("map")                | Program name
+        * "type"        | string ("test bed"/"dut")     | Type of the device
```
Having this documentation per command is really excellent.

morantr replied:
thanks :)

arnout commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -312,7 +312,92 @@ void wfa_ca::handle_wfa_ca_message(
+
+        // TODO: Configure device "name". Currently not needed.
+        // NOTE: This action is an asynchronous procedure and will be required to be executed
+        //       from task infrastructure.
```
I don't believe this comment is true - I think we can just update the database and be done with it. This gets processed from the son_master_thread which "owns" the database.

Does CAPI require the running and completed as separate responses, or can you just respond with completed right away?

morantr replied:
It is an asynchronous process since you need to config an Agent which requires to send a message to the agent and get back a response, and when all of this is done you return "complete".

And yes, the "running" is needed after validity check of the command, and on commands that it may take a lot of time to return a response - like configuring an agent.
For commands like "dev_get_param" you can return an immediate complete status since all the data is locally on the controller and it doesn't need to ask the agent to send it to him.

arnout replied:
> It is an asynchronous process since you need to config an Agent which requires to send a message to the agent and get back a response, and when all of this is done you return "complete".

 OK, then I misunderstood the meaning of this CAPI command. I thought it was meant to set the parameters locally in the controller, not to go and configure the agents. I thought the configuration would be triggered by a send_1905 autoconfig_renew or by triggering an autoconfig on the agent itself.

arnout commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -309,7 +319,136 @@ void wfa_ca::handle_wfa_ca_message(
+        if (mandatory_params_num < params.size()) {
+
+            if (!cmdu_tx.create(g_mid, ieee1905_1::eMessageType(message_type))) {
+                LOG(ERROR) << "cmdu creation of type CHANNEL_SELECTION_REQUEST_MESSAGE, has failed";
```
CHANNEL_SELECTION_REQUEST_MESSAGE -> message_type_str

morantr replied:
fixed

arnout commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -408,3 +547,60 @@ void wfa_ca::handle_wfa_ca_message(
+                    LOG(ERROR) << "addClass tlvChannelPreference has failed";
+                    return false;
+                }
+            } else if (tlv_type == int(wfa_map::eTlvTypeMap::TLV_CHANNEL_PREFERENCE)) {
```
TLV_TRANSMIT_POWER_LIMIT

morantr replied:
fixed

arnout commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -408,3 +547,60 @@ void wfa_ca::handle_wfa_ca_message(
+        tlv_type = cmdu_tx.getNextTlvType();
+        while (tlv_type != int(ieee1905_1::eTlvType::TLV_END_OF_MESSAGE)) {
+            if (tlv_type == int(wfa_map::eTlvTypeMap::TLV_CHANNEL_PREFERENCE)) {
+                if (!cmdu_tx.addClass<wfa_map::tlvChannelPreference>()) {
```
Why do you bother parsing this? You don't do anything with the channel preference TLV.

I suppose this is something to be added in a later commit? In that case, I don't see why you wouldn't also delay the parsing until that later commit.

In fact, I don't see the point of this entire function...

morantr replied:
Because I loaded preprepared buffer containing the data but the CMDU I created is new and empty. So I need to perform AddClass() to every TLV which was added to the buffer, in order to push it to CMDU classes vector.
It is necessary because when we finalize the TLV on the send function,  we iterate over this vector and do swapping.

arnout commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -408,3 +547,60 @@ void wfa_ca::handle_wfa_ca_message(
+
+    int tlv_type;
+
+    switch (ieee1905_1::eMessageType(message_type)) {
```
message_type is already an eMessageType.

morantr replied:
thanks. fixed.

arnout reviewed (approved):
There is one essential comment: the send_cmdu_to_agent of the channel preference query has to move inside the condition.

I will approve now, please fixup the nitpicks while rebasing, or leave them as is of course.

I think the approach with the certification buffer tx message is a horrible hack and shouldn't be done. I don't think it saves us much effort compared to just storing the channel preference response information in the database and doing a pseudo-channel selection based on that information. However, since we're already very far down that path, I will not oppose it.

Regarding the commit messages: I made very strong complaints about that in my earlier review. Now the review is done, it is a bit less relevant. I would still like to see an explanation why the certification buffer is needed and what it is meant for; the other explanations why are less important.

tomereli commented:
19 (!) commits - can't this be split to 2 PRs or even 3? we said we don't want such huge PRs since they make reviewing hard.

morantr commented:
> 19 (!) commits - can't this be split to 2 PRs or even 3? we said we don't want such huge PRs since they make reviewing hard.

It 10 commits. All other are fixups...

morantr committed:
framework: tlvf: recursive set string bug fix

Problem:
When the TLVF generates setter to a dynamic array of type char, it
creates two functions that get std::string as an argument.
One of them get const and the other one non-const, but it does
cast to const and then calling (hopefully) to the function that gets
const.
The setter that gets non-cost std::string is unnecessary because
the setter function doesn't change it, so it should be const anyway.
The problem that observed is that the setter that gets the non-const
the argument does not call to the other setter, and call itself
recursively, and create an infinite loop.

Solution:
Remove the unnecessary non-const setter.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr commented:
I added one more commit that fixes a bug that I found when I tested the feature:
framework: tlvf: recursive set string bug fix (4dee193)
